### ISSUE TITLE
SQL Filters get randomly reordered without any possible modification

### ISSUE THREAD

**Issue by marcoruggine** [metabase/metabase#5136]
Hi, I've run into a very strange behaviour in the SQL editor: sometimes it rearranges my filters in an order that can't be modified but in no way is the order we need/want. After some testing it appears to happen with more than 8 different filters (in my case), and always in the same weird order. 

I've copied/pasted the code in a new question to better understand and the result is the following:

- This is the random order of a saved question, you can see that Min/Max date are far apart and so are the Index ones.
<img width="2215" alt="schermata 2017-05-17 alle 16 16 34" src="https://cloud.githubusercontent.com/assets/20041428/26190166/552a7908-3ba8-11e7-8e42-0301a4a97297.png">

- If I copy the same exact code in a new question and assign the correct parameters for each filter, I have the correct order  untill I'll save the question:
<img width="2404" alt="schermata 2017-05-17 alle 16 19 26" src="https://cloud.githubusercontent.com/assets/20041428/26190173/5a852d8a-3ba8-11e7-82c5-8a33f3b69bb3.png">

- If I save the question, I get back to the order in screenshot one

How can I decide the order of my filters? What am I missing here?

I'm running V 0.24 on a Postgres database, Thanks!

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by salsakran** [comment]
Related to #4776, which should allow you to reorder widgets. 

No idea why they are being reordered on saving. @tlrobinson is there a default sort order to parameter widgets?

**Comment by attekei** [comment]
Parameters are stored in `dataset_query`  JSON like this:
```
{
    "database": 1,
    "type": "native",
    "native": {
        "query": "SELECT * FROM ORDERS WHERE A = {{A}} OR B = {{B}} OR C= {{C}}",
        "template_tags": {
            "A": {
                "id": "a680fa9d-840b-3fdd-6ef6-a88dd6908b9b",
                "name": "A",
                "display_name": "A",
                "type": "text"
            },
            "B": {
                "id": "f53ae7a2-4c00-f575-8f8c-b900afc49e1f",
                "name": "B",
                "display_name": "B",
                "type": "text"
            },
            "C": {
                "id": "f38bc472-a446-b60c-30c7-86e4c368e3a2",
                "name": "C",
                "display_name": "C",
                "type": "text"
            }
        }
    }
}
```

In JSON objects are unordered by definition, i.e. the order of `template_tags` isn't guaranteed. I think that the JSON deserialization in Clojure that breaks the order if there are more than 2^3 entities in the object. Maybe @camsaul knows more about that.

Being able to reorder dashboard widgets makes a lot of sense, but is the same needed for this context (parameters in SQL questions)?

**Comment by camsaul** [comment]
My guess is it's using an array map under the hood for smaller size maps since a linear search for a key would be faster than the overhead of hashing at smaller sizes. At some certain point it switches to a hash map instead which would explain the change in ordering.

At any rate if we want things kept in a certain order maps are not the right data type for the task. We can either store all the param maps in an array or add an additional array with param order

**Comment by attekei** [comment]
I'd vote for an additional array because that information is only relevant to frontend and doesn't require modifying existing `template_tags` logic in both backend and frontend.

**Comment by dannyeuu** [comment]
There are any future planning for this bug?

**Comment by sldorman** [comment]
We need the ability to order parameters on SQL questions.  I would like to order them based on common usage.  For example, any required parameters would be first, then optional parameters would be ordered by what the user would be most likely to filter.

**Comment by marcoruggine** [comment]
Hi guys, any update on this?

**Comment by MarcGJA** [comment]
Related to https://github.com/metabase/metabase/issues/3238.  Got the same problem, added a 9th filter to a question and POOF!  All the filters randomly re-ordered themselves!  Awe... sacrificed one and had to redo them all from scratch... :(

**Comment by chucklessmith** [comment]
Yeah, this is a pretty ugly situation, unfortunately. I get that it's an edge case (I mean what lunatic is going around adding a dozen filters to a report, besides me?) but it does just kill the usability of the question.

**Comment by mazameli** [comment]
Elevating priority of this: seems it's affecting a fair number of folks.

**Comment by AlexRizzuto** [comment]
Any update on this?

**Comment by anton-khodak** [comment]
Any update on this?

**Comment by aagz** [comment]
In v0.38.1 the problem still exists.

**Comment by aagz** [comment]
Hello. Any update on this?

**Comment by vlad-metelev** [comment]
Re: any update on this?

**Comment by unnyns-307** [comment]
Hello all, I also still face this problem on Metabase V.0.43.4. Any updates on this issue?

**Comment by AaronLhiu** [comment]
I'm suffering this problem on Metabase V0.50.19...

**Comment by ranquild** [comment]
This might be caused by clojure and not FE, to double check.

**Comment by kamilmielnik** [comment]
> This might be caused by clojure and not FE, to double check.

> I think that the JSON deserialization in Clojure that breaks the order if there are more than 2^3 entities in the object.

This is exactly what is happening.
It's `Lib.extractTemplateTags` call here: https://github.com/metabase/metabase/blob/37b4269d1bb2d11c742b98b1c91435a4ea6bd176/frontend/src/metabase-lib/v1/queries/NativeQuery.ts#L415

`Object.keys(this.templateTagsMap())` will return expected order of keys while `Object.keys(Lib.extractTemplateTags(queryText, this.templateTagsMap()))` will return shuffled order of keys.

Reassigning to BE.

**Comment by kamilmielnik** [comment]
Similar to #51614

**Comment by ixipixi** [comment]
Some additional context:
https://metaboat.slack.com/archives/C052ZBWRG3W/p1782927906781049

**Comment by camsaul** [comment]
Failing test in `metabase.lib-be.models.transforms-test`

```clj
(deftest ^:parallel preserve-template-tag-order-test
  (testing "Should preserve parameter order when deserializing MBQL from JSON (#5136, QUE2-607)"
    (let [json ((:in lib-be/transform-query)
                {:lib/type :mbql/query
                 :stages   [{:lib/type :mbql.stage/native
                             :template-tags
                             (into
                              (ordered-map/ordered-map)
                              (map (fn [i]
                                     (let [param-name (format "parameter_%d" i)]
                                       [param-name
                                        {:widget-type  :category
                                         :id           (format "00000000-0000-0000-0000-00000000000%d" i)
                                         :name         param-name
                                         :display-name (format "Parameter %d" i)
                                         :type         :dimension
                                         :dimension    [:field {} 1]
                                         :default      nil}])))
                              (range 10))
                             :native   "<<NATIVE QUERY>>"}]
                 :database 2})
          query ((:out lib-be/transform-query) json)]
      (is (= ["parameter_0"
              "parameter_1"
              "parameter_2"
              "parameter_3"
              "parameter_4"
              "parameter_5"
              "parameter_6"
              "parameter_7"
              "parameter_8"
              "parameter_9"]
             (keys (get-in query [:stages 0 :template-tags])))))))
```
